File: php_legacy/scripts/import_csv.php
<?php
declare(strict_types=1);

/**
 * JOSAA CSV Importer
 *
 * Usage:  php scripts/import_csv.php [path/to/file.csv]
 *
 * - Filters to IITs only
 * - Normalizes preparatory ranks ("1234P") and flags them
 * - Caches dim lookups in memory to avoid round-trips
 * - Batches fact inserts (1000 rows per INSERT) inside one transaction
 */

require_once __DIR__ . '/../config/Database.php';

final class JosaaImporter
{
    private PDO $pdo;
    private array $iitCache    = [];
    private array $branchCache = [];
    private array $quotaCache  = [];
    private array $seatCache   = [];
    private array $genderCache = [];

    private const BATCH_SIZE = 1000;

    private const NEW_AGE_KEYWORDS = [
        'Artificial Intelligence', 'Data Science',
        'Mathematics and Computing', 'Mathematics & Computing',
        'AI and Data',
    ];
    private const CORE_KEYWORDS = [
        'Mechanical', 'Civil', 'Chemical', 'Metallurgical',
    ];
    private const CSE_KEYWORDS = ['Computer Science'];

    private const OLD_IITS = [
        'Indian Institute of Technology Bombay',
        'Indian Institute of Technology Delhi',
        'Indian Institute of Technology Kanpur',
        'Indian Institute of Technology Kharagpur',
        'Indian Institute of Technology Madras',
        'Indian Institute of Technology Roorkee',
        'Indian Institute of Technology (BHU) Varanasi',
        'Indian Institute of Technology Guwahati',
    ];

    public function __construct(PDO $pdo)
    {
        $this->pdo = $pdo;
    }

    public function import(string $csvPath): void
    {
        if (!is_readable($csvPath)) {
            throw new RuntimeException("CSV not readable: {$csvPath}");
        }

        $fh = fopen($csvPath, 'r');
        if (!$fh) {
            throw new RuntimeException("Could not open CSV: {$csvPath}");
        }

        $header = fgetcsv($fh);
        if (!$header) {
            throw new RuntimeException("CSV appears to be empty");
        }

        $colIndex = array_flip(array_map(
            static fn($h) => strtolower(trim($h)),
            $header
        ));

        foreach (['iit','branch','quota','seat_type','gender','or','cr','year','round'] as $required) {
            if (!isset($colIndex[$required])) {
                throw new RuntimeException("Missing required column: {$required}");
            }
        }

        $this->pdo->beginTransaction();

        try {
            $batch = [];
            $rowCount = 0;
            $skipped  = 0;

            while (($row = fgetcsv($fh)) !== false) {
                $iitName = trim($row[$colIndex['iit']]);

                // Keep IITs only
                if (stripos($iitName, 'Indian Institute of Technology') === false) {
                    $skipped++;
                    continue;
                }

                [$openRank, $openPrep] = $this->normalizeRank($row[$colIndex['or']]);
                [$closeRank, $closePrep] = $this->normalizeRank($row[$colIndex['cr']]);

                if ($openRank === 0 || $closeRank === 0) {
                    $skipped++;
                    continue;
                }

                $batch[] = [
                    'iit_id'         => $this->getOrInsertIit($iitName),
                    'branch_id'      => $this->getOrInsertBranch(trim($row[$colIndex['branch']])),
                    'quota_id'       => $this->getOrInsertDim('dim_quota', 'quota_code', 'quota_id',
                                        trim($row[$colIndex['quota']]), $this->quotaCache),
                    'seat_type_id'   => $this->getOrInsertDim('dim_seat_type', 'seat_type_code', 'seat_type_id',
                                        trim($row[$colIndex['seat_type']]), $this->seatCache),
                    'gender_id'      => $this->getOrInsertDim('dim_gender', 'gender_code', 'gender_id',
                                        trim($row[$colIndex['gender']]), $this->genderCache),
                    'year'           => (int) $row[$colIndex['year']],
                    'round_no'       => (int) $row[$colIndex['round']],
                    'opening_rank'   => $openRank,
                    'closing_rank'   => $closeRank,
                    'is_preparatory' => ($openPrep || $closePrep) ? 1 : 0,
                ];

                if (count($batch) >= self::BATCH_SIZE) {
                    $this->flushBatch($batch);
                    $rowCount += count($batch);
                    $batch = [];
                    echo "Inserted {$rowCount} rows...\n";
                }
            }

            if (!empty($batch)) {
                $this->flushBatch($batch);
                $rowCount += count($batch);
            }

            $this->pdo->commit();
            fclose($fh);

            // Delete non-last round data as requested
            $this->pdo->exec("
                DELETE f1 FROM fact_allotment f1
                LEFT JOIN (
                    SELECT year, MAX(round_no) as max_round
                    FROM fact_allotment
                    GROUP BY year
                ) f2 ON f1.year = f2.year AND f1.round_no = f2.max_round
                WHERE f2.max_round IS NULL
            ");

            echo "\n✔ Import complete.\n";
            echo "  Rows inserted: {$rowCount}\n";
            echo "  Rows skipped:  {$skipped}\n";
        } catch (Throwable $e) {
            $this->pdo->rollBack();
            fclose($fh);
            throw $e;
        }
    }

    /**
     * Returns [int $rank, bool $isPreparatory].
     * A rank suffixed with "P" (e.g. "1234P") is a preparatory rank.
     */
    private function normalizeRank(string $raw): array
    {
        $raw = trim($raw);
        $prep = (bool) preg_match('/P/i', $raw);
        $rank = (int) preg_replace('/[^0-9]/', '', $raw);
        return [$rank, $prep];
    }

    private function getOrInsertIit(string $name): int
    {
        if (isset($this->iitCache[$name])) {
            return $this->iitCache[$name];
        }

        $stmt = $this->pdo->prepare("SELECT iit_id FROM dim_iit WHERE iit_name = ?");
        $stmt->execute([$name]);
        $id = $stmt->fetchColumn();

        if (!$id) {
            $generation = in_array($name, self::OLD_IITS, true) ? 'old' : 'new';
            $shortCode  = $this->deriveShortCode($name);
            $ins = $this->pdo->prepare(
                "INSERT INTO dim_iit (iit_name, short_code, generation) VALUES (?, ?, ?)"
            );
            $ins->execute([$name, $shortCode, $generation]);
            $id = (int) $this->pdo->lastInsertId();
        }

        return $this->iitCache[$name] = (int) $id;
    }

    private function deriveShortCode(string $fullName): string
    {
        $clean = preg_replace('/Indian Institute of Technology/i', 'IIT', $fullName);
        return trim(preg_replace('/\s+/', ' ', $clean ?? $fullName));
    }

    private function getOrInsertBranch(string $name): int
    {
        if (isset($this->branchCache[$name])) {
            return $this->branchCache[$name];
        }

        $stmt = $this->pdo->prepare("SELECT branch_id FROM dim_branch WHERE branch_name = ?");
        $stmt->execute([$name]);
        $id = $stmt->fetchColumn();

        if (!$id) {
            $category = $this->classifyBranch($name);
            $ins = $this->pdo->prepare(
                "INSERT INTO dim_branch (branch_name, category) VALUES (?, ?)"
            );
            $ins->execute([$name, $category]);
            $id = (int) $this->pdo->lastInsertId();
        }

        return $this->branchCache[$name] = (int) $id;
    }

    private function classifyBranch(string $name): string
    {
        foreach (self::CSE_KEYWORDS as $kw) {
            if (stripos($name, $kw) !== false) return 'cse_family';
        }
        foreach (self::NEW_AGE_KEYWORDS as $kw) {
            if (stripos($name, $kw) !== false) return 'new_age';
        }
        foreach (self::CORE_KEYWORDS as $kw) {
            if (stripos($name, $kw) !== false) return 'core';
        }
        return 'other';
    }

    private function getOrInsertDim(
        string $table,
        string $col,
        string $idCol,
        string $val,
        array &$cache
    ): int {
        if (isset($cache[$val])) {
            return $cache[$val];
        }

        $stmt = $this->pdo->prepare("SELECT {$idCol} FROM {$table} WHERE {$col} = ?");
        $stmt->execute([$val]);
        $id = $stmt->fetchColumn();

        if (!$id) {
            $ins = $this->pdo->prepare("INSERT INTO {$table} ({$col}) VALUES (?)");
            $ins->execute([$val]);
            $id = (int) $this->pdo->lastInsertId();
        }

        return $cache[$val] = (int) $id;
    }

    private function flushBatch(array $batch): void
    {
        $placeholders = implode(',', array_fill(0, count($batch), '(?,?,?,?,?,?,?,?,?,?)'));
        $sql = "INSERT INTO fact_allotment
                (iit_id, branch_id, quota_id, seat_type_id, gender_id,
                 year, round_no, opening_rank, closing_rank, is_preparatory)
                VALUES {$placeholders}";

        $flat = [];
        foreach ($batch as $r) {
            array_push($flat,
                $r['iit_id'], $r['branch_id'], $r['quota_id'], $r['seat_type_id'],
                $r['gender_id'], $r['year'], $r['round_no'],
                $r['opening_rank'], $r['closing_rank'], $r['is_preparatory']
            );
        }

        $this->pdo->prepare($sql)->execute($flat);
    }
}

// -------- CLI Bootstrap --------
// Only auto-run when called directly from the command line.
// When included by public/import.php (browser version), this block is skipped.
if (PHP_SAPI === 'cli' && isset($argv) && realpath($argv[0]) === __FILE__) {
    $csvPath = $argv[1] ?? __DIR__ . '/../data/josaa_2016_2022.csv';

    try {
        $pdo = Database::connect();
        (new JosaaImporter($pdo))->import($csvPath);
    } catch (Throwable $e) {
        fwrite(STDERR, "❌ Import failed: " . $e->getMessage() . PHP_EOL);
        exit(1);
    }
}

File: analytics/management/commands/import_csv.py
"""
JOSAA CSV importer — Django management command.

Loads a JoSAA seat-allotment CSV into the star schema. (This is the ETL step
that follows the Selenium scrape; the scraper itself is maintained separately.)

What it does:
  - keeps IITs only (drops NITs/IIITs/GFTIs)
  - normalizes preparatory ranks ("1234P") and flags them via is_preparatory
  - pre-classifies branch category (cse_family/new_age/core/other) and IIT
    generation (old/new) at load time, so analytical queries filter on clean
    enums instead of pattern-matching free text
  - caches dimension lookups in memory to avoid per-row round-trips
  - batches fact inserts inside a single transaction
  - keeps EVERY round by default (use --final-round-only to keep just the final round per year)

Usage:
    python manage.py import_csv [path/to/file.csv] [--batch-size 1000] [--final-round-only]

Expected CSV columns (case-insensitive):
    id, iit, branch, quota, seat_type, gender, or, cr, year, round
"""
import csv
import re
from pathlib import Path

from django.conf import settings
from django.core.management.base import BaseCommand, CommandError
from django.db import connection, transaction

OLD_IITS = {
    'Indian Institute of Technology Bombay',
    'Indian Institute of Technology Delhi',
    'Indian Institute of Technology Kanpur',
    'Indian Institute of Technology Kharagpur',
    'Indian Institute of Technology Madras',
    'Indian Institute of Technology Roorkee',
    'Indian Institute of Technology (BHU) Varanasi',
    'Indian Institute of Technology Guwahati',
}
CSE_KEYWORDS = ['Computer Science']
NEW_AGE_KEYWORDS = [
    'Artificial Intelligence', 'Data Science',
    'Mathematics and Computing', 'Mathematics & Computing', 'AI and Data',
]
CORE_KEYWORDS = ['Mechanical', 'Civil', 'Chemical', 'Metallurgical']

REQUIRED_COLUMNS = ['iit', 'branch', 'quota', 'seat_type', 'gender', 'or', 'cr', 'year', 'round']


class Command(BaseCommand):
    help = 'Import a JoSAA seat-allotment CSV into the star schema.'

    def add_arguments(self, parser):
        parser.add_argument(
            'csv_path', nargs='?',
            default=str(Path(settings.BASE_DIR) / 'data' / 'josaa_2016_2022.csv'),
            help='Path to the JoSAA CSV (default: data/josaa_2016_2022.csv)',
        )
        parser.add_argument('--batch-size', type=int, default=1000,
                            help='Rows per fact INSERT (default: 1000)')
        parser.add_argument('--final-round-only', action='store_true',
                            help='Keep only the final round per year (default: keep every round)')

    def handle(self, *args, **opts):
        importer = JosaaImporter(
            self.stdout, self.style,
            batch_size=opts['batch_size'],
            final_round_only=opts['final_round_only'],
        )
        importer.run(opts['csv_path'])


class JosaaImporter:
    def __init__(self, stdout, style, batch_size=1000, final_round_only=False):
        self.out = stdout
        self.style = style
        self.batch_size = batch_size
        self.final_round_only = final_round_only
        self._iit, self._branch = {}, {}
        self._quota, self._seat, self._gender = {}, {}, {}

    def run(self, csv_path):
        path = Path(csv_path)
        if not path.is_file():
            raise CommandError(f'CSV not found: {csv_path}')

        with path.open(newline='', encoding='utf-8') as fh, transaction.atomic():
            reader = csv.reader(fh)
            try:
                header = next(reader)
            except StopIteration:
                raise CommandError('CSV appears to be empty')

            col = {h.strip().lower(): i for i, h in enumerate(header)}
            missing = [c for c in REQUIRED_COLUMNS if c not in col]
            if missing:
                raise CommandError(f'Missing required column(s): {", ".join(missing)}')

            batch, inserted, skipped = [], 0, 0
            for row in reader:
                if not row:
                    continue

                iit_name = row[col['iit']].strip()
                if 'indian institute of technology' not in iit_name.lower():
                    skipped += 1
                    continue

                open_rank, open_prep = self._norm_rank(row[col['or']])
                close_rank, close_prep = self._norm_rank(row[col['cr']])
                if open_rank == 0 or close_rank == 0:
                    skipped += 1
                    continue

                batch.append((
                    self._get_iit(iit_name),
                    self._get_branch(row[col['branch']].strip()),
                    self._get_dim('dim_quota', 'quota_code', 'quota_id', row[col['quota']].strip(), self._quota),
                    self._get_dim('dim_seat_type', 'seat_type_code', 'seat_type_id', row[col['seat_type']].strip(), self._seat),
                    self._get_dim('dim_gender', 'gender_code', 'gender_id', row[col['gender']].strip(), self._gender),
                    int(row[col['year']]),
                    int(row[col['round']]),
                    open_rank, close_rank,
                    1 if (open_prep or close_prep) else 0,
                ))

                if len(batch) >= self.batch_size:
                    self._flush(batch)
                    inserted += len(batch)
                    batch = []
                    self.out.write(f'Inserted {inserted} rows...')

            if batch:
                self._flush(batch)
                inserted += len(batch)

            if self.final_round_only:
                self._delete_non_final_rounds()

        self.out.write(self.style.SUCCESS(
            f'\nImport complete. Rows inserted: {inserted}, rows skipped: {skipped}'
        ))

    # ── rank normalization ───────────────────────────────────────────────
    @staticmethod
    def _norm_rank(raw):
        """Return (int rank, bool is_preparatory). A 'P' suffix marks a prep rank."""
        raw = (raw or '').strip()
        prep = bool(re.search('P', raw, re.IGNORECASE))
        digits = re.sub(r'[^0-9]', '', raw)
        return (int(digits) if digits else 0, prep)

    # ── dimension lookups (memoized) ─────────────────────────────────────
    def _get_iit(self, name):
        if name in self._iit:
            return self._iit[name]
        with connection.cursor() as c:
            c.execute('SELECT iit_id FROM dim_iit WHERE iit_name = %s', [name])
            row = c.fetchone()
            if row:
                iid = row[0]
            else:
                generation = 'old' if name in OLD_IITS else 'new'
                short = re.sub(r'\s+', ' ', re.sub(r'Indian Institute of Technology', 'IIT', name, flags=re.IGNORECASE)).strip()
                c.execute('INSERT INTO dim_iit (iit_name, short_code, generation) VALUES (%s, %s, %s)',
                          [name, short, generation])
                iid = c.lastrowid
        self._iit[name] = iid
        return iid

    def _get_branch(self, name):
        if name in self._branch:
            return self._branch[name]
        with connection.cursor() as c:
            c.execute('SELECT branch_id FROM dim_branch WHERE branch_name = %s', [name])
            row = c.fetchone()
            if row:
                bid = row[0]
            else:
                c.execute('INSERT INTO dim_branch (branch_name, category) VALUES (%s, %s)',
                          [name, self._classify(name)])
                bid = c.lastrowid
        self._branch[name] = bid
        return bid

    @staticmethod
    def _classify(name):
        low = name.lower()
        if any(k.lower() in low for k in CSE_KEYWORDS):
            return 'cse_family'
        if any(k.lower() in low for k in NEW_AGE_KEYWORDS):
            return 'new_age'
        if any(k.lower() in low for k in CORE_KEYWORDS):
            return 'core'
        return 'other'

    def _get_dim(self, table, col, id_col, val, cache):
        # table/col/id_col are internal constants (never user input) — safe to interpolate.
        if val in cache:
            return cache[val]
        with connection.cursor() as c:
            c.execute(f'SELECT {id_col} FROM {table} WHERE {col} = %s', [val])
            row = c.fetchone()
            if row:
                did = row[0]
            else:
                c.execute(f'INSERT INTO {table} ({col}) VALUES (%s)', [val])
                did = c.lastrowid
        cache[val] = did
        return did

    # ── bulk fact insert ─────────────────────────────────────────────────
    @staticmethod
    def _flush(batch):
        placeholders = ','.join(['(%s,%s,%s,%s,%s,%s,%s,%s,%s,%s)'] * len(batch))
        sql = (
            'INSERT INTO fact_allotment '
            '(iit_id, branch_id, quota_id, seat_type_id, gender_id, '
            'year, round_no, opening_rank, closing_rank, is_preparatory) '
            f'VALUES {placeholders}'
        )
        flat = [v for rec in batch for v in rec]
        with connection.cursor() as c:
            c.execute(sql, flat)

    @staticmethod
    def _delete_non_final_rounds():
        with connection.cursor() as c:
            c.execute(
                """
                DELETE f1 FROM fact_allotment f1
                LEFT JOIN (
                    SELECT year, MAX(round_no) AS max_round
                    FROM fact_allotment
                    GROUP BY year
                ) f2 ON f1.year = f2.year AND f1.round_no = f2.max_round
                WHERE f2.max_round IS NULL
                """
            )
